src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -372,7 +372,106 @@ protected void InitializeReferences()
+                runner.runtext = input;
+                ReadOnlySpan<char> span = input.AsSpan(beginning, length);
+                runner.InitializeForScan(this, span, startat - beginning, quick);
+
+                int stoppos = RightToLeft ? 0 : span.Length;
+
+                // If previous match was empty or failed, advance by one before matching.
+                if (prevlen == 0)
+                {
+                    if (runner.runtextstart == stoppos)
+                    {
+                        return RegularExpressions.Match.Empty;
+                    }
+
+                    runner.runtextpos += RightToLeft ? -1 : 1;
+                }
+
+                return InternalPerformScan(quick, input, beginning, runner, span, returnNullIfQuick: true);
+            }
+            finally
+            {
+                _runner = runner;
+            }
+        }
+
+        private static Match? InternalPerformScan(bool quick, string input, int beginning, RegexRunner runner, ReadOnlySpan<char> span, bool returnNullIfQuick)
+        {
+            // We need to save the original beginning of the input in case we are in the CompileToAssembly case, which relies on
+            // setting runtextbeg and runtextend correctly. We cannot calculate these values on the fly based on input and the span
+            // because it is not guaranteed that span.Length + beginning = input.
+            runner.originalRuntextbeg = beginning;
+            runner.Scan(span);
+
+            Match? match = runner.runmatch;
+            // if we got a match, set runmatch to null if quick is true
+            if (match!.FoundMatch)
+            {
+                runner.runtext = null; // drop reference to text to avoid keeping it alive in a cache
+
+                if (match.Text != input)
+                {
+                    match.Text = input;
+                }
+
+                if (quick && returnNullIfQuick)
+                {
+                    runner.runmatch!.Text = null; // Drop reference to text
+                    return null;
+                }
+
+                if (!quick)
+                    runner.runmatch = null;
+
+                match.Tidy(runner.runtextpos);
+
+                // If there was a match and the original text was sliced, then add beginning to the index to get the real
+                // Index of the match.
+                if (match.Success && beginning != 0)
+                {
+                    match.AddBeginningToIndex(beginning);
+                }
+
+                return match;
+            }
+
+            runner.runtext = null; // drop reference to text to avoid keeping it alive in a cache
+
+            if (!quick)
+            {
+                runner.runmatch = null;
+            }
+            else
+            {
+
+                if (runner.runmatch != null)
+                {
+                    runner.runmatch.Text = null;
+                }
```

Comment:
Why would runner.runmatch be null here?  Should this just be:
```suggestion
                match.Text = null;
```
?